pkg/fanal/artifact/image/image.go

```diff
@@ -213,22 +219,46 @@ func (a Artifact) inspect(ctx context.Context, missingImage string, layerKeys, b
 
 		layerInfo, err := a.inspectLayer(ctx, layer, disabledAnalyzers)
 		if err != nil {
-			return nil, xerrors.Errorf("failed to analyze layer (%s): %w", layer.DiffID, err)
-		}
-		if err = a.cache.PutBlob(layerKey, layerInfo); err != nil {
-			return nil, xerrors.Errorf("failed to store layer: %s in cache: %w", layerKey, err)
+			return Layer{}, xerrors.Errorf("failed to analyze layer (%s): %w", layer.DiffID, err)
 		}
-		if lo.IsNotEmpty(layerInfo.OS) {
-			osFound = layerInfo.OS
+
+		return Layer{
+			Key:  layerKey,
+			Info: layerInfo,
+		}, nil
+	}, func(layer Layer) error {
+		layers = append(layers, layer)
+		if lo.IsNotEmpty(layer.Info.OS) {
+			osFound = layer.Info.OS
 		}
-		return nil, nil
 
-	}, nil)
+		return nil
+	})
 
 	if err := p.Do(ctx); err != nil {
 		return xerrors.Errorf("pipeline error: %w", err)
 	}
 
+	p2 := parallel.NewPipeline(a.artifactOption.Parallel, false, layers, func(ctx context.Context, layer Layer) (Layer, error) {
+		// Call post handlers to modify layer info
+		layer.Info.OS.Merge(osFound)
+		if err := a.handlerManager.PostHandle(ctx, nil, &layer.Info); err != nil {
+			return Layer{}, xerrors.Errorf("post handler error: %w", err)
+		}
+
+		return layer, nil
+	}, func(layer Layer) error {
+		if err := a.cache.PutBlob(layer.Key, layer.Info); err != nil {
+			return xerrors.Errorf("failed to store layer: %s in cache: %w", layer.Key, err)
+		}
+
+		return nil
+	})
```

Comment:
I'm not sure if one layer should affect another layer before saving to cache.
I can't think of an example right now, but my intuition tells me we don't need to do this.

I think it's better to update info after we take all layers from cache.
@knqyf263 can you share your opinion if you have time?

Reply:
No problem on that, we can keep pushing the original layer to the cache by adding this small change:
```diff
diff --git a/pkg/fanal/artifact/image/image.go b/pkg/fanal/artifact/image/image.go
index b9775a4d..880e42e3 100644
--- a/pkg/fanal/artifact/image/image.go
+++ b/pkg/fanal/artifact/image/image.go
@@ -241,8 +241,9 @@ func (a Artifact) inspect(ctx context.Context, missingImage string, layerKeys, b

        p2 := parallel.NewPipeline(a.artifactOption.Parallel, false, layers, func(ctx context.Context, layer Layer) (Layer, error) {
                // Call post handlers to modify layer info
-               layer.Info.OS.Merge(osFound)
-               if err := a.handlerManager.PostHandle(ctx, nil, &layer.Info); err != nil {
+               layerInfo := layer.Info
+               layerInfo.OS.Merge(osFound)
+               if err := a.handlerManager.PostHandle(ctx, nil, &layerInfo); err != nil {
                        return Layer{}, xerrors.Errorf("post handler error: %w", err)
                }
```

Reply:
1 more possible way - https://github.com/aquasecurity/trivy/pull/5439#discussion_r1423464277

Reply:
done at https://github.com/aquasecurity/trivy/pull/5439/commits/d2222197d294178711b19d065aaebaf976b73b37